presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    public long getFileSize()
+    {
+        return fileSize;
+    }
+
+    public Slice getTailSlice()
+    {
+        return Slices.wrappedBuffer(tailBuffer, 0, length);
+    }
+
+    public FSDataInputStream replaceStreamWithContentsIfComplete(FSDataInputStream originalStream)
+            throws IOException
+    {
+        if (fileSize == length) {
+            originalStream.close();
+            return new FSDataInputStream(new ByteArraySeekablePositionedReadableInputStream(tailBuffer, 0, length));
+        }
+        return originalStream;
+    }
+
+    @VisibleForTesting
+    static final class ByteArraySeekablePositionedReadableInputStream
+            extends ByteArrayInputStream
+            implements Seekable, PositionedReadable
+    {
+        private ByteArraySeekablePositionedReadableInputStream(byte[] array, int offset, int length)
+        {
+            super(array, offset, length);
+        }
+
+        @Override
+        public void seek(long pos)
+                throws IOException
+        {
+            if (pos < 0) {
+                throw new IOException(FSExceptionMessages.NEGATIVE_SEEK);
+            }
+            if (pos > count) {
+                throw new IOException(FSExceptionMessages.CANNOT_SEEK_PAST_EOF);
+            }
+            this.reset();
+            this.skip(pos);
+        }
+
+        @Override
+        public long getPos()
+        {
+            return pos;
+        }
+
+        @Override
+        public boolean seekToNewSource(long targetPos)
+        {
+            return false;
+        }
+
+        /**
+         * Implementation taken directly from {@link org.apache.hadoop.fs.FSInputStream}
+         */
+        @Override
+        public int read(long position, byte[] buffer, int offset, int length)
+                throws IOException
+        {
+            synchronized (this) {
+                long oldPos = getPos();
+                int nread = -1;
+                try {
+                    seek(position);
+                    nread = read(buffer, offset, length);
+                }
+                finally {
+                    seek(oldPos);
+                }
+                return nread;
+            }
+        }
+
+        /**
+         * Implementation taken directly from {@link org.apache.hadoop.fs.FSInputStream}
+         */
+        @Override
+        public void readFully(long position, byte[] buffer, int offset, int length)
+                throws IOException
+        {
+            int nread = 0;
+            while (nread < length) {
+                int nbytes = read(position + nread, buffer, offset + nread, length - nread);
+                if (nbytes < 0) {
+                    throw new EOFException("End of file reached before reading fully.");
+                }
+                nread += nbytes;
+            }
+        }
+
+        /**
+         * Implementation taken directly from {@link org.apache.hadoop.fs.FSInputStream}
```

Comment:
No need for the comment since this is the obvious way to implement this method (as in basically every `InputStream`).